core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -531,8 +531,22 @@ assert equalType(
               final UnifyResult result = rule.apply(call);
               if (result != null) {
                 ++count;
-                attempted.add(new Replacement(result.call.query, result.result));
-                MutableRel parent = result.call.query.replaceInParent(result.result);
+                final MutableRel result1;
+                if (targetDescendant == target
+                    && !result.result.rowType.equals(result.call.query.rowType)) {
+                  // When target and query match with each other but have different
+                  // rowTypes, we add a MutableProject for compensation, thus to
+                  // assure the rowType of the replaced is exactly as before.
+                  final RexBuilder rexBuilder = target.cluster.getRexBuilder();
+                  List<RexNode> projects = new ArrayList<>(
+                      rexBuilder.identityProjects(result.call.query.rowType));
+                  result1 = MutableProject.of(
+                      result.call.query.rowType, result.result, projects);
+                } else {
+                  result1 = result.result;
```

Comment:
Instead of new a MutableProject, can we just check the project list of the result row type ? If i'm not mistaken, only if the query and target have same row expressions but different alias can be matched, maybe we should only check the expression for alias.

Reply:
Sorry, I don't quite get it ~ Do you mean below ?
1. No need to new a `MutableProject`, though the row type is different, we just let it go
2. Add a check/assert for the row type, we allow name mismatch, but other parts of row type should match with each other

Reply:
> Sorry, I don't quite get it ~ Do you mean below ?
> 
> 1. No need to new a `MutableProject`, though the row type is different, we just let it go
> 2. Add a check/assert for the row type, we allow name mismatch, but other parts of row type should match with each other

Yeah, that is just what i mean.